modeshape-jcr/src/main/java/org/modeshape/jcr/spi/index/provider/IndexChangeAdapters.java

```diff
@@ -334,26 +439,35 @@ protected void removeNode( String workspaceName,
                                    Path path,
                                    Name primaryType,
                                    Set<Name> mixinTypes ) {
-            if (includeRoot || path.isRoot()) {
-                index.remove(nodeKey(key));
-            }
+            index().remove(nodeKey(key));
         }
 
         @Override
-        protected void completeChanges() {
-            index.commit();
-            super.completeChanges();
-        }
-
-        @Override
-        protected void completeWorkspaceChanges() {
-            index.commit();
-            super.completeWorkspaceChanges();
+        protected void renameNode( String workspaceName, 
+                                   NodeKey key, 
+                                   Path newPath, 
+                                   Path.Segment oldSegment, 
+                                   Name primaryType,
+                                   Set<Name> mixinTypes ) {
+            PathFactory pathFactory = context.getValueFactories().getPathFactory();
+            Path oldPath = pathFactory.create(newPath.subpath(0, newPath.size()), oldSegment);
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
         }
 
         @Override
-        public String toString() {
-            return getClass().getSimpleName() + "(\"" + index.getName() + "\")";
+        protected void reorderNode( String workspaceName, 
+                                    NodeKey key, 
+                                    Name primaryType, 
+                                    Set<Name> mixinTypes, 
+                                    NodeKey parent,
+                                    Path newPath, 
+                                    Path oldPath, 
+                                    Path reorderedBeforePath ) {
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
```

Comment:
First, I *think* you can use the `oldPath` and `newPath` to check the SNS index of the node that was reordered. If the SNS index is `1` in both places, then that should imply that no other nodes were affected (right?), and that the existing logic is sufficient.

However, I do think we have to address what happens when the SNS indexes are different. But, first we need to make sure that no other events are generated that correspond to the other changes. That would be ideal, because those other events would be handled. (But I don't think there are other events, though it may be easier to generate additional events for those cases. Something to consider.)

But, if the SNS indexes are different and no other events handle the other affected nodes, then we probably can rely upon the index provider to help. For example, it can search its index to find all nodes that have the same path (except different SNS indexes). The question is whether to add a method that does this, or to use queries to do this. Either way, this change adapter can then use the results to tell the index to change the other affected nodes. 

And of course, if we add a method on the index provider and we want to keep it backward compatible, we'd want to have a default implementation. If the default implementation did nothing (essentially saying no other nodes are affected by this SNS change), then the index could become inconsistent. However, if the default implementation could return a result that said it doesn't know (e.g., null rather than a collection or iterator), then can the change adapter cause a reindexing of just the children of the parent (and not descendants of the children)? That's not ideal, but at least it would keep the index from becoming inconsistent; and once a provider implements the method correctly, then it becomes efficient again.

WDYT?